### PR TITLE
Introduce gci tool

### PR DESCRIPTION
**Pull request by daixiang0**
Signed-off-by: Xiang Dai <long0dai@foxmail.com>

Details refer to [this](https://github.com/istio/istio/pull/25597#issuecomment-662203597).

### CODE REVIEW

**Comment by howardjohn** [comment]
I am a bit concerned this is relying on saw string processing rather than smarter code from https://github.com/golang/tools/tree/master/internal/imports. And how this will impact things like duplicate imports, comments on imports, etc

**Comment by daixiang0** [comment]
It runs after golangci-lint and ensure the import orders.

 `goimport` in golangci-lint would check deplicate and sort imports, but it can not ensure the import order  since it can not handle empty lines like [issue](https://github.com/istio/istio/issues/25129) said.

**Comment by daixiang0** [comment]
ping @howardjohn @carolynhu

**Comment by carolynhu** [comment]
@daixiang0 Thanks for providing this tool, do you have some run examples? I think it's better to add those examples in your README file. The cases you handled and results you gonna show.

**Comment by daixiang0** [comment]
@carolynhu done.

**Comment by howardjohn** [comment]
I am concerned by having a tool like this that is doing raw text processing. I would think something based on goimports is safer. Like https://github.com/golang/tools/pull/150/files, go.coder.com/go-tools/cmd/goimports

**Comment by howardjohn** [comment]
https://github.com/kamilsk/go-tools/releases/tag/goimports works well, just tried it out

**Comment by daixiang0** [comment]
`goimport` would generate AST and then do format while `gci` do format without AST, i think it is fast as well.

And i try this, but do not works as expected with comment, but `gci` does:
```
root:[istio]$ goimports -ungroup -d pkg/envoy/proxy.go
diff -u pkg/envoy/proxy.go.orig pkg/envoy/proxy.go
--- pkg/envoy/proxy.go.orig	2020-07-24 14:08:15.154433601 +0800
+++ pkg/envoy/proxy.go	2020-07-24 14:08:15.154433601 +0800
@@ -22,17 +22,14 @@
 	"os/exec"
 	"path"
 	"strings"
-	"time"
+	"time" // assafa

-       // assafa
 	envoyAdmin "github.com/envoyproxy/go-control-plane/envoy/admin/v3"
 	"github.com/gogo/protobuf/types"
-
 	meshconfig "istio.io/api/mesh/v1alpha1"
+	"istio.io/istio/pkg/bootstrap"
 	"istio.io/pkg/env"
 	"istio.io/pkg/log"
-
-	"istio.io/istio/pkg/bootstrap"
 )

 const (
root:[istio]$ gci -d -local istio pkg/envoy/proxy.go
diff -u pkg/envoy/proxy.go.orig pkg/envoy/proxy.go
--- pkg/envoy/proxy.go.orig	2020-07-24 14:08:32.138559944 +0800
+++ pkg/envoy/proxy.go	2020-07-24 14:08:32.138559944 +0800
@@ -24,15 +24,14 @@
 	"strings"
 	"time"

-       // assafa
+	// assafa
 	envoyAdmin "github.com/envoyproxy/go-control-plane/envoy/admin/v3"
 	"github.com/gogo/protobuf/types"

 	meshconfig "istio.io/api/mesh/v1alpha1"
+	"istio.io/istio/pkg/bootstrap"
 	"istio.io/pkg/env"
 	"istio.io/pkg/log"
-
-	"istio.io/istio/pkg/bootstrap"
 )

 const (

```

**Comment by carolynhu** [comment]
parse AST is necessary when formatting the code. I think goimports helps with sorting imports, but not grouping them by certain groups in order. A simple/basic raw string processing in this use case would be fine if we just want those imports get grouped in an expected order and group. This does not touch code parsing part, so I do not see anything wrong this would cause.

File: cmd/gci/README.md [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+)
+```
+
+### with comment and alias
+
+```go
+package main
+import (
+  "fmt"
+  _ "github.com/golang" // golang
```
Comment (carolynhu) [inline cmd/gci/README.md:67]: could you add both `inline` comment and `above line` comment examples?

**Comment by mandarjog** [comment]
https://github.com/istio/istio/blob/release-0.8/bin/fmt.sh#L36

This is what we used to do before:
TLDR;
1. remove blank lines from import blocks
2. run gofmt
3. run goimports

gofmt followed by goimport ensured that everything was fine.

**Comment by howardjohn** [comment]
> This is what we used to do before:

Any idea why we removed it?

**Comment by mandarjog** [comment]
Someone thought it was not necessary: https://github.com/istio/istio/pull/6798   and we should not remove spaces.

**Comment by howardjohn** [comment]
Ok, seems like the author did not realize the impact it would have. So I
guess there are really 3 options: preprocess goimports, postprocess
goimports, or fork goimports. I don't really have a strong preference as
long as it is reliable and fast

**Comment by daixiang0** [comment]
`gci` foucs on grouping and it runs after `goimport` which cover in `golangci-lint`, so it would not use AST.

File: cmd/gci/README.md (hunk 1 above)
Comment (daixiang0) [reply cmd/gci/README.md:67]: I have added `above line` comment, could you share `inline` comment, i am confused about it.

Comment (carolynhu) [reply cmd/gci/README.md:67]: i think you already have it: `_ "github.com/golang" // golang` here

Comment (daixiang0) [reply cmd/gci/README.md:67]: Now all done :)

**Comment by fpesce** [comment]
@mandarjog @howardjohn : Why don't we go back to what was done before, that seems like the minimal thing to do to achieve the result we want. Also relies on mature tools such as goimports/gofmt.

We could revisit this question in next T&R group.
@brian-avery / @ericvn would you mind having a look at Mandar's comment in above thread?

**Comment by daixiang0** [comment]
I think the older one does not meet current goal.

**Comment by daixiang0** [comment]
@howardjohn WDYT?

**Comment by howardjohn** [comment]
I think since this is in golangci-lint now we can probably just pull it in from there. We just updated to golangci-lint 1.30 which has it
